src/legacy/ui/public/chrome/api/nav.ts

```diff
@@ -0,0 +1,163 @@
+          LegacyNavLinkProperties
+        ]
+    )
+  );
+
+  /**
+   * Clear last url for deleted saved objects to avoid loading pages with "Could not locate..."
+   */
+  chrome.untrackNavLinksForDeletedSavedObjects = (deletedIds: string[]) => {
+    function urlContainsDeletedId(url: string) {
+      const includedId = deletedIds.find(deletedId => {
+        return url.includes(deletedId);
+      });
+      if (includedId === undefined) {
+        return false;
+      }
+      return true;
+    }
+
+    coreNavLinks.getAll().forEach(link => {
+      const legacyProps = legacyNavProps.get(link.id)!;
+      if (legacyProps.linkToLastSubUrl && urlContainsDeletedId(link.url!)) {
+        setLastUrl(link, link.appUrl);
+      }
+    });
+  };
+
+  /**
+   * Manually sets the last url for the given app. The last url for a given app is updated automatically during
+   * normal page navigation, so this should only need to be called to insert a last url that was not actually
+   * navigated to. For instance, when saving an object and redirecting to another page, the last url of the app
+   * should be the saved instance, but because of the redirect to a different page (e.g. `Save and Add to Dashboard`
+   * on visualize tab), it won't be tracked automatically and will need to be inserted manually. See
+   * https://github.com/elastic/kibana/pull/11932 for more background on why this was added.
+   *
+   * @param id {String} - an id that represents the navigation link.
+   * @param kibanaParsedUrl {KibanaParsedUrl} the url to track
+   */
+  chrome.trackSubUrlForApp = (id: string, kibanaParsedUrl: KibanaParsedUrl) => {
+    const navLink = coreNavLinks.get(id);
+    if (navLink) {
+      setLastUrl(navLink, kibanaParsedUrl.getAbsoluteUrl());
+    }
+  };
+
+  internals.trackPossibleSubUrl = async function(url: string) {
+    const kibanaParsedUrl = absoluteToParsedUrl(url, chrome.getBasePath());
+
+    for (let link of coreNavLinks.getAll()) {
+      const subUrlBase = legacyNavProps.get(link.id)!.subUrlBase;
+      const active = kibanaParsedUrl.getAppRootPath().startsWith(subUrlBase);
+      link = coreNavLinks.update(link.id, { active })!;
+
+      if (active) {
+        setLastUrl(link, url);
+        continue;
+      }
+
+      link = refreshLastUrl(link);
+
+      const newGlobalState = kibanaParsedUrl.getGlobalState();
+      if (newGlobalState) {
+        injectNewGlobalState(link, kibanaParsedUrl.appId, newGlobalState);
+      }
+    }
+  };
+
+  function lastSubUrlKey(link: ChromeNavLink) {
+    return `lastSubUrl:${link.appUrl}`;
+  }
+
+  function getLastUrl(link: ChromeNavLink) {
+    return internals.appUrlStore.getItem(lastSubUrlKey(link));
+  }
+
+  function setLastUrl(link: ChromeNavLink, url: string) {
+    internals.appUrlStore.setItem(lastSubUrlKey(link), url);
+    refreshLastUrl(link);
+  }
+
+  function refreshLastUrl(link: ChromeNavLink) {
+    const lastSubUrl = getLastUrl(link);
+
+    return coreNavLinks.update(link.id, {
```

Comment:
The interface only allows a few fields to be modified, and it does updates by creating new objects rather than modifying the existing ones. The fields that are allowed to be modified are currently only being enforced by TypeScript, I'll change that to be enforced at runtime as well and make sure these objects are frozen.

Reply:
Is this meant to be intentionally mutable? Have we considered any issues that may arise from mutating external objects, or is that the intent?